core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -1556,6 +1681,133 @@ public void warmUp() {
+        if (!ff.exists(other2)) {
+            // Backups and older versions of detached partitions will not have _meta
+            LOG.info().$("detached _meta not found, skipping check [path=").$(other2).I$();
+            return;
+        }
+        try {
+            detachedMetadata.deferredInit(other2, ColumnType.VERSION);
+            if (metadata.getId() != detachedMetadata.getId()) {
+                throw CairoException.detachedMetadataMismatch("table_id");
+            }
+            // check column name, type and isIndexed
+            for (int colIdx = 0; colIdx < columnCount; colIdx++) {
+                String columnName = metadata.getColumnName(colIdx);
+                int detColIdx = detachedMetadata.getColumnIndexQuiet(columnName);
+                if (detColIdx == -1) {
+                    throw CairoException.detachedMetadataMismatch("missing_column: " + columnName);
+                }
+                if (metadata.getColumnType(colIdx) != detachedMetadata.getColumnType(detColIdx)) {
+                    throw CairoException.detachedColumnMetadataMismatch(colIdx, columnName, "type");
+                }
+                if (metadata.isColumnIndexed(colIdx) != detachedMetadata.isColumnIndexed(detColIdx)) {
+                    throw CairoException.detachedColumnMetadataMismatch(colIdx, columnName, "is_indexed");
+                }
+            }
+            // structural changes mean columns may have been removed
+            if (metadata.getStructureVersion() != detachedMetadata.getStructureVersion()) {
+                for (int i = 0, limit = detachedMetadata.getColumnCount(); i < limit; i++) {
+                    String columnName = detachedMetadata.getColumnName(i);
+                    if (metadata.getColumnIndexQuiet(columnName) == -1) {
+                        ff.walk(detachedPath, (pUtf8NameZ, type) -> {
+                            if (type == Files.DT_FILE) {
+                                fileNameSink.clear();
+                                Chars.utf8DecodeZ(pUtf8NameZ, fileNameSink);
+                                int len = fileNameSink.length();
+                                int dotIdx = len - 1;
+                                while (dotIdx > 0 && fileNameSink.charAt(dotIdx) != '.') {
+                                    dotIdx--;
+                                }
+                                if (dotIdx > 0) {
+                                    CharSequence cn = fileNameSink.subSequence(0, dotIdx);
+                                    if (Chars.equals(cn, columnName)) {
+                                        other2.parent().concat(fileNameSink).$();
+                                        if (!ff.remove(other2)) {
+                                            LOG.info().$("failed to remove [path=").$(other2).I$();
+                                        }
+                                    }
+                                }
+                            }
+                        });
```

Comment:
What if new columns added since the detachment happened? These should have column tops added so that are treated as nulls.

Reply:
indeed, done